runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java

```diff
@@ -0,0 +1,98 @@
+import java.util.UUID;
+
+/**
+ * Globally Unique Identity (GUID) generator that returns ids
+ * with weak comparable ordering with minimal distributed state sync.
+ * On startup increment a globally unique id via Corfu Object.
+ * Use this id with locally incrementing values to return unique ids.
+ * Higher bits are made of an UTC timestamp while lower bits carry the
+ * counter.
+ *
+ * Created by Sundar Sridharan on 5/22/19.
+ */
+@Slf4j
+public class CorfuGuidGenerator implements OrderedGuidGenerator {
+    private final static long TIMESTAMP_CORRECTION_MASK = 0xFF00000000000000L;
+    private final static int TIMESTAMP_CORRECTION_SHIFT = 56;
+    private final static long TIMESTAMP_MASK = 0x00FFffFFffFF0000L;
+    private final static int TIMESTAMP_SHIFT = 16;
+    private final static int MAX_TIMESTAMP_CORRECTION = 0xFF;
+    private final static int MAX_INSTANCE_ID = 0xFFff;
+
+    private final String GUID_STREAM_NAME = "CORFU_GUID_COUNTER_STREAM";
+
+    private CorfuCounter distributedCounter;
+
+    private long timestampCorrection = 0L;
+    private Long previousTimestamp = 0L;
+    private long instanceId;
+
+    public CorfuGuidGenerator(CorfuRuntime rt) {
+        distributedCounter = rt.getObjectsView().build()
+                .setType(CorfuCounter.class)
+                .setStreamName(GUID_STREAM_NAME)
+                .open();
+        updateInstanceId();
+    }
+
+    private long updateInstanceId() {
+        long nextInstanceId = distributedCounter.next();
+        instanceId = nextInstanceId;
+        return nextInstanceId;
+    }
+
+    /**
+     * +----------------------+---------------+--------------------+
+     * | Timestamp Correction | UTC Timestamp | Unique Instance ID |
+     * +----------------------+---------------+--------------------+
+     * <------1 byte ---------><--5 bytes----><------2 bytes------->
+     *
+     * @return a globally unique id with minimal distributed sync up
+     */
+    @Override
+    public long nextLong() {
+        long currentTimestamp = System.currentTimeMillis();
+        if (currentTimestamp <= previousTimestamp) {
+            timestampCorrection++;
+            if (timestampCorrection > MAX_TIMESTAMP_CORRECTION) {
+                timestampCorrection = 0;
+                updateInstanceId();
+            }
+        }
+        previousTimestamp = currentTimestamp;
+        return (timestampCorrection << TIMESTAMP_CORRECTION_SHIFT) |
+                ((currentTimestamp << TIMESTAMP_SHIFT) & TIMESTAMP_MASK) |
+                instanceId & MAX_INSTANCE_ID;
+    }
+
+    /**
+     * +------------------------------+------------------------------+
+     * | Correction | Timestamp       |      Instance Id             |
+     * +------------------------------+------------------------------+
+     * <----1 byte--><---7 bytes-----><-------- 8 bytes ------------->
+     *
+     * @return non lossy UUID carrying both epoch and sequence
+     */
+    @Override
+    public UUID nextUUID() {
```

Comment:
From what i understand, all this machinery is needed to generate unique keys, but is this all needed?
Why isn't it just <instanceId><counter>, where instanceId is a sequencer number obtained from the sequencer and counter is a local atomic long that gets incremented on every generation.

Reply:
1. With timestamp baked in, it allows debugging since the key carries the time when it was generated.
2. Given that typical deployments have NTP, this also brings in global ordering across nodes.
I thought these are good properties to have?